File: LEVEL3/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    queue<pair<string, int>> que;
    vector<int> chk(words.size(), 0);
    
    que.push({ begin, 0 });
    
    while(!que.empty()) {
        string start = que.front().first;
        int cnt = que.front().second;
        
        que.pop();
        
        for(int i = 0; i < words.size(); i++) {
            int k = 0;
            
            if(chk[i] != 0) {
                continue;
            }
            
            for(int j = 0; j < words[i].size(); j++) {
                if(start[j] != words[i][j]) {
                    k++;
                }
            }
            
            if(k == 1) {
                if(words[i] == target) {
                    return cnt + 1;
                }
                
                chk[i] = true;
                que.push({ words[i], cnt + 1 });
            }
        }
    }
    
    return answer;
}

File: LEVEL3/단어 변환.java
import java.io.*;
import java.util.*;

class Solution {
    static boolean visited[];
    static int answer = 987654321;
    
    public int solution(String begin, String target, String[] words) {
        for(int i = 0; i < words.length; i++) {
            if(count(begin, words[i]) <= 1) {
                visited = new boolean[words.length];
                visited[i] = true;
                dfs(1, i, target, words);
                
            }
            
        }
        return answer == 987654321 ? 0 : answer;
    }
    
    static int count(String s1, String s2) {
        int cnt = 0;
        
        for(int i = 0; i < s1.length(); i++) {
            if(s1.charAt(i) != s2.charAt(i)) {
                cnt++;
            }
        }
        
        return cnt;
    }
    
    static void dfs(int cnt, int cur, String target, String[] words) {
        if(target.equals(words[cur])) {
            answer = Math.min(cnt, answer);
            return;
        }
        
        for(int i = 0; i < words.length; i++) {
            if(!visited[i] && count(words[cur], words[i]) == 1) {
                visited[i] = true;
                dfs(cnt + 1, i, target, words);
                visited[i] = false;   
            }
        }
    }
}